### PR TITLE
Do not automatically add -ObjC flag when integrating Objective-C dependencies

### PR DESCRIPTION
**Pull request by thedavidharris**
This reverts commit 92970222bfa7a0e5fab9a56a86623907d375c4f5.

Resolves https://github.com/tuist/tuist/issues/6243

### Short description 📝

This reverts a change that automatically places `-ObjC` compiler flags with no opt-out behavior. This can cause issues in projects and make them unbuildable or bloat their binary size unnecessarily.

### How to test the changes locally 🧐

Test with `app_with_spm_dependencies` fixture and see that the `-ObjC` flag is not automatically added

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint:fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Review by fortmarek — COMMENTED** [review]
I am aligned with reverting the change but then we need to add good docs around this. I would add a section noting what to do when someone runs into the runtime error somewhere in [this](https://docs.tuist.io/guide/project/dependencies.html#external-dependencies) page. Would you mind adding that as a part of this PR?

Tagging @freak4pc, so he's aware of this change.

**Comment by freak4pc** [comment]
> I am aligned with reverting the change but then we need to add good docs around this. I would add a section noting what to do when someone runs into the runtime error somewhere in [this](https://docs.tuist.io/guide/project/dependencies.html#external-dependencies) page. Would you mind adding that as a part of this PR?
> 
> 
> 
> Tagging @freak4pc, so he's aware of this change.

Thanks 🙏 we'll try to add manually where needed. Still seems to me Apple has some other built in behavior in SPM that makes things work even without this flag, so pretty interesting what that could be

**Comment by thedavidharris** [comment]
@fortmarek absolutely, coming back to this today

@freak4pc can you confirm which of your dependencies are running into this and see if any of the other resolutions mentioned in https://github.com/tuist/tuist/issues/6243 work? Happy to chat over Slack as well to see if we can identify any other patterns/resoloutions.

I am definitely curious as to why this seems to work with SPM but Apple definitely does some funky linking things there. Yet, some docs (like Firebase) seemed to say it _was_ needed for SPM. 

I do think the best long term strategy is for Tuist to support something similar to Bazel's `alwayslink` that then uses `force-load` flags, although this is definitely trickier.

**Comment by freak4pc** [comment]
> @freak4pc can you confirm which of your dependencies are running into this and see if any of the other resolutions mentioned in #6243 work? Happy to chat over Slack as well to see if we can identify any other patterns/resoloutions.

The issued was with GoogleSignIn, IMO. I'll see if we have some time to debug this anytime soon.

File: docs/docs/guide/project/dependencies.md [hunk 1]
```diff
@@ -194,3 +194,11 @@ pod install
...
 > [!WARNING] 
 > CocoaPods dependencies are not compatible with workflows like `build` or `test` that run `xcodebuild` right after generating the project. They are also incompatible with binary caching and selective testing since the fingerprinting logic doesn't account for the Pods dependencies.
+
+## Objective-C Dependencies
+
+When integrating Objective-C dependencies, the inclusion of certain flags on the consuming target may be necessary to avoid runtime crashes as detailed in [Apple Technical Q&A QA1490](https://developer.apple.com/library/archive/qa/qa1490/_index.html).
+
+Since the build system and Tuist have no way of inferring whether the flag is necessary or not, and since the flag comes with potentially undesirable side effects, Tuist will not automatically apply any of these flags, and because Swift Package Manager considers `-ObjC` to be included via an `.unsafeFlag` most packages cannot include it as part of their default linking settings when required.
+
+Consumers of Objective-C dependencies (or internal Objective-C targets) should apply  `-ObjC` or `-force_load` flags when required by setting `OTHER_LDFLAGS` on consuming targets.
```
Comment (fortmarek) [inline docs/docs/guide/project/dependencies.md:204]: The `app_with_spm_dependencies` fixture requires the `-ObjC` flag. You can double check that by running the app on the simulator. Let's add it and we can link to it as an example through references, once a README is added: https://docs.tuist.io/reference/examples/app_with_build_rules.html

File: fixtures/app_with_spm_dependencies/App/Sources/AppKit/AppKit.swift [hunk 2]
```diff
@@ -32,7 +32,7 @@ public enum AppKit {
         _ = YAMLEncoder()
 
         // Use GoogleSignIn
-        _ = GIDSignIn.sharedInstance.hasPreviousSignIn()
+        _ = GIDSignIn.sharedInstance.configuration
```
Comment (fortmarek) [inline fixtures/app_with_spm_dependencies/App/Sources/AppKit/AppKit.swift:35]: I'd keep using `hasPreviousSignIn()` that requires the `-ObjC` flag as an exxample. Or why was this changed?

File: docs/docs/guide/project/dependencies.md (hunk 1 above)
Comment (fortmarek) [reply docs/docs/guide/project/dependencies.md:204]: @thedavidharris once this is resolved, we can merge this 🙂

Comment (thedavidharris) [reply docs/docs/guide/project/dependencies.md:204]: Added!

Comment (fortmarek) [reply docs/docs/guide/project/dependencies.md:204]: Thanks! However, the documentation here still misses a link to the fixture 🙏
